hedera-node/hedera-app/src/main/java/com/hedera/node/app/records/files/StreamFileProducerBase.java

```diff
@@ -0,0 +1,355 @@
+ * <p>
+ *     Threading model: is really important here, This class depends on the fact that public methods are all called on
+ *     handle transaction thread. So are single threaded and executed in the correct order.
+ * </p>
+ * <p>
+ *     This base class has no mutable state, all methods should be stateless so they can be called from any thread in subclasses.
+ * </p>
+ */
+@SuppressWarnings("SwitchStatementWithTooFewBranches")
+public abstract class StreamFileProducerBase implements AutoCloseable {
+    protected static final Logger log = LogManager.getLogger(StreamFileProducerBase.class);
+    /** The file extension for record files */
+    public static final String RECORD_EXTENSION = "rcd";
+    /** The suffix added to RECORD_EXTENSION when they are compressed */
+    public static final String COMPRESSION_ALGORITHM_EXTENSION = ".gz";
+
+    /** signer for generating signatures */
+    private final Signer signer;
+    /** The HAPI protobuf version */
+    protected final SemanticVersion hapiVersion;
+    /** The version of the record file format */
+    protected final int recordFileVersion;
+    /** The version of the signature file format */
+    private final int signatureFileVersion;
+    /** The maximum size of a sidecar file in bytes */
+    private final int maxSideCarSizeInBytes;
+    /** The directory where record files are written */
+    protected final Path nodeScopedRecordLogDir;
+    /** The directory where sidecar files are written */
+    private final Path nodeScopedSidecarDir;
+    /** True if record files should be compressed on creation */
+    protected final boolean compressFilesOnCreation;
+    /** The format for creating the record files */
+    protected final RecordFileFormat recordFileFormat;
+
+    /**
+     * Construct RecordManager and start background thread
+     *
+     * @param configProvider the configuration to read from
+     * @param nodeInfo the current node information
+     * @param signer the signer to use for signing in signature files
+     * @param fileSystem the file system to use, needed for testing to be able to use a non-standard file
+     *                   system. If null default is used.
+     */
+    public StreamFileProducerBase(
+            @NonNull final ConfigProvider configProvider,
+            @NonNull final NodeInfo nodeInfo,
+            @NonNull final Signer signer,
+            @Nullable final FileSystem fileSystem) {
+        this.signer = signer;
+        // read configuration
+        BlockRecordStreamConfig recordStreamConfig =
+                configProvider.getConfiguration().getConfigData(BlockRecordStreamConfig.class);
+        recordFileVersion = recordStreamConfig.recordFileVersion();
+        signatureFileVersion = recordStreamConfig.signatureFileVersion();
+        compressFilesOnCreation = recordStreamConfig.compressFilesOnCreation();
+        maxSideCarSizeInBytes = recordStreamConfig.sidecarMaxSizeMb() * 1024 * 1024;
```

Comment:
Here we're loading the `BlockRecordStreamConfig` from a `ConfigProvider`, and then reading the fields and saving them in final fields on the instance. A couple of thoughts:

- We could just save the `BlockRecordStreamConfig` instance. It is a record, and looking up values on it should be inexpensive? It saves from having lots of redundant fields on this class.
- If we're going to read and save all that config up front, then we could pass a `Configuration` instead of a `ConfigProvider`. The purpose of the provider is to give a way to get the current config at some arbitrary point in time, whereas if we're reading all the state in the constructor, we don't really benefit from a config
- I'm not sure if the config for the record stream stuff should be dynamic or static. Here it is static. We should decide which way to go on this.